sockeye/transformer.py

```diff
@@ -139,57 +158,89 @@ def __call__(self,
+                                  source,
+                                  source_lengths,
+                                  source_max_length)
+        target = self.post_attention(target, prev, target_max_length)
+
+        # feed-forward
+        prev = self.ff(self.pre_ff(target, None, target_max_length),
+                       target_max_length)
+        target = self.post_ff(target, prev, target_max_length)
+
         return target
 
 
-class TransformerResidual:
+class TransformerProcessBlock:
     """
-    Residual connection with optional layer normalization.
+    Block to perform pre/post processing on layer inputs.
+    The processing steps are determined by the sequence argument, which can contain one of the three operations:
+    n: layer normalization
+    r: residual connection
+    d: dropout
     """
 
     def __init__(self,
+                 sequence: str,
                  num_hidden: int,
-                 layer_normalization: bool,
                  dropout: float,
                  prefix: str) -> None:
+        self.sequence = sequence
         self.num_hidden = num_hidden
         self.dropout = dropout
-        self.layer_norm = None
         self.prefix = prefix
-        if layer_normalization:
-            self.layer_norm = layers.LayerNormalization(num_hidden=self.num_hidden, prefix=self.prefix)
+        self.layer_norm = None
+        if "n" in sequence:
+            self.layer_norm = layers.LayerNormalization(num_hidden=self.num_hidden, prefix="%snorm" % self.prefix)
 
-    def __call__(self, x: mx.sym.Symbol, y: mx.sym.Symbol, length: int) -> mx.sym.Symbol:
+    def __call__(self,
+                 data: mx.sym.Symbol,
+                 prev: Optional[mx.sym.Symbol],
+                 length: int) -> mx.sym.Symbol:
         """
-        Apply residual connections with optional layer normalization and dropout.
+        Apply processing sequence to data with optional previous input.
 
-        :param x: (batch, length, num_hidden).
-        :param y: (batch, length, num_hidden).
-        :param length: maximum sequence length.
-        :return: (batch, length, num_hidden).
+        :param data: Input data. Shape: (batch, length, num_hidden).
+        :param prev: Previous data. Shape: (batch, length, num_hidden).
+        :param length: Maximum sequence length.
+        :return: Processed data. Shape: (batch, length, num_hidden).
         """
-        if self.dropout > 0.0:
-            y = mx.sym.Dropout(y, p=self.dropout)
-        z = x + y
-        if self.layer_norm is not None:
-            z = self._reshape_and_normalize(z, length)
-        return z
+        if not self.sequence:
+            return data
+
+        if prev is None:
+            assert 'r' not in self.sequence, "Residual connection not allowed if no previous value given."
+
+        for step in self.sequence:
+
+            if step == "r":
+                data = mx.sym._internal._plus(data, prev, name="%sresidual" % self.prefix)
```

Comment:
Can't we use mxnet's residual implementation?

Reply:
What residual implementation are you referring to, thats not an RNN cell :)
Its just the '+' operator but I am using the internal plus to be able to name it.

Reply:
Right, residual in mxnet is for RNNs.